### PR TITLE
Convert bool to error, helper func for cd to skew

### PR DESCRIPTION
**Pull request by fejta**
Switch from various functions returning `False` when something bad happens to returning an `error`
Encapsulate logic to switch to the skew directory inside chdirSkew

Also add a TODO for using `hyphen-flags` instead of `underscore_flags`

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30321)

### CODE REVIEW

**Comment by fejta** [comment]
Can a brother get a code review @ixdy @krousey?

File: hack/e2e.go [hunk 1]
```diff
@@ -443,10 +446,15 @@ func finishRunning(stepName string, cmd *exec.Cmd) bool {
...
-		log.Printf("Error running %v: %v", stepName, err)
-		return false
+		return fmt.Errorf("Error running %v: %v", stepName, err)
+	}
+	return nil
+}
+
+func doOrDie(msg string, err error) {
+	if err != nil {
+		log.Fatalf("error %s: %s", msg, err)
```
Comment (krousey) [inline hack/e2e.go]: Please don't do this. I know it sucks to repeat code, but please `log.Fatal` where the error is received and not in a utility method like this. What you have here will always die on line 456. https://play.golang.org/p/SdbuTBrh8P

File: hack/e2e.go [hunk 2]
```diff
@@ -78,41 +86,36 @@ func main() {
...
-		status := 1
-		if IsUp() {
-			status = 0
-			log.Printf("Cluster is UP")
-		} else {
-			log.Printf("Cluster is DOWN")
+		if !IsUp() {
+			log.Fatal("Cluster is DOWN")
 		}
-		os.Exit(status)
```
Comment (krousey) [inline hack/e2e.go]: Is this behavior change intentional? Do we no longer want to exit here?

File: hack/e2e.go [hunk 3]
```diff
@@ -443,10 +446,15 @@ func finishRunning(stepName string, cmd *exec.Cmd) bool {
 	}(time.Now())
 
 	if err := cmd.Run(); err != nil {
-		log.Printf("Error running %v: %v", stepName, err)
-		return false
+		return fmt.Errorf("Error running %v: %v", stepName, err)
```
Comment (krousey) [inline hack/e2e.go]: Error messages shouldn't start with capital letters.

File: hack/e2e.go [hunk 4]
```diff
@@ -126,109 +129,116 @@ func main() {
...
-	if !success {
-		os.Exit(1)
+	if errs != nil {
+		log.Fatalf("Encountered %d errors: %v", len(errs), errs)
 	}
 }
 
-func writeOrDie(dir, name, data string) string {
-	f, err := os.Create(filepath.Join(dir, name))
+func writeOrDie(dir, name, data string) (string, error) {
```
Comment (krousey) [inline hack/e2e.go]: `writeOrDie` seems like a misnomer now. Also, it's now equivalent to

```go
ioutil.WriteFile(filepath.Join(dir, name), []byte(data))
```

File: hack/e2e.go [hunk 5]
```diff
@@ -126,109 +129,116 @@ func main() {
...
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			errs = appendError(errs, DiffResources(beforeResources, upResources, afterResources, *dump))
+		}
 	}
 
-	if !success {
-		os.Exit(1)
+	if errs != nil {
```
Comment (krousey) [inline hack/e2e.go]: Since this is a slice `if len(errs) != 0` might be better. It works with both nil and empty slices.

File: hack/e2e.go (hunk 1 above)
Comment (fejta) [inline hack/e2e.go]: Done

File: hack/e2e.go (hunk 2 above)
Comment (fejta) [inline hack/e2e.go]: Remove --isup entirely

File: hack/e2e.go (hunk 3 above)
Comment (fejta) [inline hack/e2e.go]: Done

File: hack/e2e.go (hunk 4 above)
Comment (fejta) [inline hack/e2e.go]: Done

File: hack/e2e.go (hunk 5 above)
Comment (fejta) [inline hack/e2e.go]: Done

**Comment by fejta** [comment]
@krousey thanks for the comments. PTAL when you have the chance!

**Comment by fejta** [comment]
@k8s-bot unit test this issue #30947

**Comment by fejta** [comment]
@k8s-bot unit test this issue #IGNORE

**Comment by fejta** [comment]
@k8s-bot e2e test this issue #28275

**Comment by krousey** [comment]
A couple more minor nits. Feel free to fix and add lgtm.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[hack/e2e.go, line 202 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30321#-KPoZFstoop41f-Hbkcm:-KPoZFsuS5dytiiSs7jY:b-azl0bv) ([raw file](https://github.com/kubernetes/kubernetes/blob/c871db42d1a2156cf5824ad8fe34e5ebc69a6e94/hack/e2e.go#L202)):*
> ```Go
> 	ap := filepath.Join(location, "gcp-resources-after.txt")
> 	var mode os.FileMode
> 	mode = 0664
> ```

This assignment can be on the same line. `var mode os.FileMode = 0644`

---

*[hack/e2e.go, line 207 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30321#-KPoZdD1usk7AdTDKJP9:-KPoZdD2rUT6Xb0EZ9nz:bc44r0j) ([raw file](https://github.com/kubernetes/kubernetes/blob/c871db42d1a2156cf5824ad8fe34e5ebc69a6e94/hack/e2e.go#L207)):*
> ```Go
> 		return err
> 	}
> 	if err := ioutil.WriteFile(filepath.Join(location, "gcp-resources-cluster-up.txt"), clusterUp, mode); err != nil {
> ```

maybe a variable for this path too?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30321#-:-KPoZtumQXQtN5gsL-1x:bb8hp5u)*

**Comment by fejta** [comment]
Thanks!

---

Review status: 1 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[hack/e2e.go, line 202 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30321#-KPoZFstoop41f-Hbkcm:-KPt-hERBFRSwMghD_Tk:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c871db42d1a2156cf5824ad8fe34e5ebc69a6e94/hack/e2e.go#L202)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> This assignment can be on the same line. `var mode os.FileMode = 0644`
</details>
Done.

---

*[hack/e2e.go, line 207 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30321#-KPoZdD1usk7AdTDKJP9:-KPt-hqEdJsm1tXwLpP9:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/c871db42d1a2156cf5824ad8fe34e5ebc69a6e94/hack/e2e.go#L207)):*
<details><summary><i>Previously, krousey wrote…</i></summary>
> maybe a variable for this path too?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30321#-:-KPt-k3HRlXLk39NsaBJ:b-ucfknh)*

**Comment by fejta** [comment]
@k8s-bot e2e test this issue #30687

**Comment by fejta** [comment]
@k8s-bot test this issue #27462

**Comment by fejta** [comment]
@k8s-bot gke e2e test this issue #27462

File: hack/e2e.go [hunk 6]
```diff
@@ -126,112 +127,106 @@ func main() {
...
-			log.Fatal(err)
+			return fmt.Errorf("Could not create e2e-check-resources temp dir: %s", err)
 		}
 	}
-	bp := writeOrDie(location, "gcp-resources-before.txt", before)
-	writeOrDie(location, "gcp-resources-cluster-up.txt", clusterUp)
-	ap := writeOrDie(location, "gcp-resources-after.txt", after)
+
+	var mode os.FileMode = 0664
+	bp := filepath.Join(location, "gcp-resources-before.txt")
```
Comment (ixdy) [inline hack/e2e.go]: you have taken the Go unnecessarily-short-variable-name style to heart, I see. :)

File: hack/e2e.go [hunk 7]
```diff
@@ -126,112 +127,106 @@ func main() {
...
 		if err != nil {
-			log.Fatal(err)
+			return fmt.Errorf("Could not create e2e-check-resources temp dir: %s", err)
 		}
 	}
-	bp := writeOrDie(location, "gcp-resources-before.txt", before)
-	writeOrDie(location, "gcp-resources-cluster-up.txt", clusterUp)
-	ap := writeOrDie(location, "gcp-resources-after.txt", after)
+
+	var mode os.FileMode = 0664
```
Comment (ixdy) [inline hack/e2e.go]: nit: this could be `const`?

File: hack/e2e.go [hunk 8]
```diff
@@ -422,7 +407,7 @@ func SkewTest() bool {
 			fmt.Sprintf("--check_version_skew=%t", *checkVersionSkew)))
 }
 
-func Test() bool {
+func Test() error {
```
Comment (ixdy) [inline hack/e2e.go]: there's a `ValidateClusterSize` call in here that still `Fatal`'s on errors.

File: hack/e2e.go (hunk 7 above)
Comment (fejta) [inline hack/e2e.go]: Ack

File: hack/e2e.go (hunk 6 above)
Comment (fejta) [inline hack/e2e.go]: Ack

File: hack/e2e.go (hunk 8 above)
Comment (fejta) [inline hack/e2e.go]: Done

**Comment by fejta** [comment]
@k8s-bot node e2e test this issue #31633

**Comment by fejta** [comment]
Adding lgtm per comments and retest-not-required as we do not use the head version of hack/e2e.go in testing.

**Pull request merged**